### PR TITLE
Add a `Receiver.close()` method

### LINKED ISSUE #297: A way to explicitly stop a `Receiver`
If you stop reading from a channel, you will get an overflow warning message after a short while. 

Sometimes, however we want to properly stop and not get spammed with log warnings, so a convenient and explicit way to do that would be nice.

### PR DESCRIPTION
**Pull request by shsms**
This method would allow individual receivers to be closed, without affecting the underlying channel, if there is one.

### CODE REVIEW

**Comment by shsms** [comment]
Still need to add tests.

File: src/frequenz/channels/_merge.py [hunk 1]
```diff
@@ -187,6 +191,21 @@ def consume(self) -> ReceiverMessageT_co:
...
+    @override
+    def close(self) -> None:
+        """Close the receiver.
+
+        After calling this method, new messages will not be received.  Once the
+        receiver's buffer is drained, trying to receive a message will raise a
+        [`ReceiverStoppedError`][frequenz.channels.ReceiverStoppedError].
+        """
+        for task in self._pending:
+            if not task.done() and task.get_loop().is_running():
```
Comment (llucax) [inline src/frequenz/channels/_merge.py:203]: Mmm, interesting the `task.get_loop().is_running()`. Was this to address warnings in tests or any other reason? I wonder if it is a good practice to do in general or not (and for a normal cleanup all tasks should be cancelled before stopping the loop).

**Review by llucax — COMMENTED** [review]
LGTM but the CI is failing, and this reminds me that it would be good to add tests at least for the receivers that actually do something when `close()` is called.

And oh, yes! As usual names, names, names :laughing: 

TL;DR: All good the name actually, just some other comments about future directions.

I was going to suggest to rename `close()` to `stop()`, for once to match the `ReceiverStoppedError` but also to match the upcoming [`Service` interface](https://github.com/frequenz-floss/frequenz-core-python/blob/v1.x.x/src/frequenz/core/asyncio/_service.py).

But I see close actually wait for the current buffer to be consumed, my idea for `stop()` would be that stuff gets just dropped.

Maybe to make it play better with `Service` we can keep `close()` as you implemented it, as a "graceful" shutdown, and when we introduce `Service` we can use `cancel()` to drop the current buffers. Then maybe `stop()` can take an option to decide if `close()` or `cancel()` is called (stop is usually implemented as `self.cancel(); await self`).

**Comment by shsms** [comment]
> LGTM but the CI is failing, and this reminds me that it would be good to add tests at least for the receivers that actually do something when `close()` is called.

Yes, I'm already working on tests in this same branch, just made an early PR because the code is ready for review.

> Maybe to make it play better with `Service` we can keep `close()` as you implemented it, as a "graceful" shutdown, and when we introduce `Service` we can use `cancel()` to drop the current buffers. Then maybe `stop()` can take an option to decide if `close()` or `cancel()` is called (stop is usually implemented as `self.cancel(); await self`).

I think `stop()` doesn't make sense for receivers, because it is not a task.  Idiomatically, we should have `stop()` only for things that look like there must be an activity there.  So it is great for `Actor`, etc.  But for a stream, I like `close()` better.

File: src/frequenz/channels/_merge.py (hunk 1 above)
Comment (shsms) [reply src/frequenz/channels/_merge.py:203]: > Was this to address warnings in tests or any other reason?

Yes, I think I started adding it just for tests and I'm sure we'll find it in other places in this repo as well (unless you removed it).

> I wonder if it is a good practice to do in general or not (and for a normal cleanup all tasks should be cancelled before stopping the loop).

I agree, these could go away once we reach a point where removing them won't cause warning spam.

**Comment by llucax** [comment]
> I think stop() doesn't make sense for receivers, because it is not a task. Idiomatically, we should have stop() only for things that look like there must be an activity there. So it is great for Actor, etc. But for a stream, I like close() better.

Mmm, receiver don't **need** to have task, but many **do** (like `Merger`, `Timer` and `FileWatcher`). If we don't add a task-like interface, it means we can't stop these receivers using the `Receiver` interface, which is the only way to access them if we apply a `filter` or `map` to them for example.

But maybe there are other solutions to this. We can discuss this when the time comes.

**Comment by llucax** [comment]
Please let me know when this is ready for a final review :)

**Comment by shsms** [comment]
> Mmm, receiver don't **need** to have task, but many **do** (like `Merger`, `Timer` and `FileWatcher`). If we don't add a task-like interface, it means we can't stop these receivers using the `Receiver` interface, which is the only way to access them if we apply a `filter` or `map` to them for example.

That doesn't sound like an issue, because people still call `close()` on the receiver, and if the underlying thing is a `FileWatcher` for example, the `close()` would just call the `stop()`, which is what this PR is doing already.

And calling `close()` on `Map` and `Filter` would call `close()` of the underlying receiver.  This PR has that too.

**Comment by llucax** [comment]
> That doesn't sound like an issue, because people still call close() on the receiver, and if the underlying thing is a FileWatcher for example, the close() would just call the stop(), which is what this PR is doing already.

But `stop()` is async, and it is important to be able to wait until the thing actually stopped to avoid those annoying warnings in test we were talking about :laughing: 

If we make `close()` async, then fine, but then we are ACKing that receivers can have background running tasks, and then why shouldn't they follow the `Service` interface?

**Comment by shsms** [comment]
This is ready, the only big change is that `Receiver.close()` is now async.

**Comment by llucax** [comment]
Last push to have a (`Background`)`Service`-compatible interface.

> If we make `close()` async, then fine, but then we are ACKing that receivers can have background running tasks, and then why shouldn't they follow the `Service` interface?

**Comment by llucax** [comment]
BTW, the method can't be abstract, otherwise this is a breaking change. We need to provide a default empty implementation I guess. We can create an issue to make the method abstract for 2.0.

File: src/frequenz/channels/file_watcher.py [hunk 2]
```diff
@@ -229,6 +232,11 @@ def consume(self) -> Event:
         change, path_str = self._changes.pop()
         return Event(type=EventType(change), path=pathlib.Path(path_str))
 
+    @override
+    async def close(self) -> None:
+        """Close this receiver."""
+        self._stop_event.set()
```
Comment (llucax) [inline src/frequenz/channels/file_watcher.py:238]: We should probably do a `await self._awatch.aclose()` here to make sure all the threads from the awatch are stopped too. Maybe even mark the receiver somehow as stopped so `ready()` doesn ´t try to use `self._awatch` again? Not sure what would happen otherwise, what kind of exception would be raised.

File: src/frequenz/channels/_anycast.py [hunk 3]
```diff
@@ -441,6 +455,14 @@ def consume(self) -> _T:
 
         return next_val
 
+    @override
+    async def close(self) -> None:
+        """Close this receiver.
+
+        After closing, the receiver will not be able to receive any more messages.
+        """
+        self._closed = True
```
Comment (llucax) [inline src/frequenz/channels/_anycast.py:464]: I think here we should also break the `ready()` if it is waiting for the condition variable. I guess this would mean adding a an event and then `async.wait()` on the first one to complete in `ready()`?

I guess this is actually the difference between *closing* and *stopping*. Close is basically analogous to `Task.cancel()`, it is like it schedules the stopping. And it is actually probably a good idea to have it separated, but we are still missing the `await` then, making `stop()` async but then not actually awaiting for anything seems weird to me.

**Review by llucax — COMMENTED** [review]
I didn't review everything, because I think the review will change a lot depending on what we do with `close` vs `stop()` (`stop()` being `close()`/`cancel()` + `__await__`).

`cancel()` and `__await__` are actually the 2 primitives used by (`Background`)`Service` too, based on `Task`, and the more I look at it, the more I think receivers should follow this interface too.

**Comment by shsms** [comment]
I've made the `close()` method not abstract and not async.  With these,
- it is no longer  a breaking change
- it becomes a signal to close(), in keeping with the idioms of the asyncio library, which was the behaviour already even though it was async, and not a "ensure it is closed before returning".

With this the two above concerns should also go away.

File: src/frequenz/channels/file_watcher.py (hunk 2 above)
Comment (shsms) [reply src/frequenz/channels/file_watcher.py:238]: switched back to non-async `close()`.

File: src/frequenz/channels/_anycast.py (hunk 3 above)
Comment (shsms) [reply src/frequenz/channels/_anycast.py:464]: switched back to non-async `close()`.

**Comment by shsms** [comment]
Rebased on latest

**Comment by shsms** [comment]
@llucax could you take a look at this again?

File: src/frequenz/channels/_receiver.py [hunk 4]
```diff
@@ -217,6 +217,15 @@ def consume(self) -> ReceiverMessageT_co:
...
         """
 
+    def close(self) -> None:
+        """Close the receiver.
+
+        After calling this method, new messages will not be available from the receiver.
+        Once the receiver's buffer is drained, trying to receive a message will raise a
+        [`ReceiverStoppedError`][frequenz.channels.ReceiverStoppedError].
+        """
+        raise NotImplementedError("close() must be implemented by subclasses")
```
Comment (llucax) [inline src/frequenz/channels/_receiver.py:227]: So we don't forget:
* https://github.com/frequenz-floss/frequenz-channels-python/issues/365

**Review by llucax — APPROVED** [review]
